### PR TITLE
[Kernel] Separate Triton Attention Kernel Launches for Prefill and Decode for FULL CUDA Graph mode

### PR DESCRIPTION
**Pull request by jvlunteren**
## Purpose

This pull request introduces separate Triton attention kernel launches for prefill and decode operations within mixed prefill/decode batches when the CUDA Graph mode is set to FULL. Splitting the launches enables tuning the parameters for prefill and decode independently, improving flexibility and potential performance in CUDA graph mode.

When the CUDA Graph mode is not FULL, the existing approach is preserved: a single attention kernel launch for the entire batch. This avoids additional kernel launch overhead, which would not be (partially) compensated by CUDA Graphs in these modes.

~~The changes in this PR are not intended to deliver significant performance gains immediately, but rather to lay the groundwork for future optimizations in upcoming PRs.~~

## Performance
The following results were obtained for `meta-llama/Llama-3.1-8B-Instruct` on an NVIDIA H100 GPU, by running

<pre>$ VLLM_ATTENTION_BACKEND=TRITON_ATTN vllm bench latency \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --input-len &ltinput-length&gt --output-len 4 \
    --batch-size 1</pre>

for
1. the current upstream version of the Triton attention kernel, and
2. the updated Triton attention kernel (this PR)

both for the default CUDA Graph mode and for FULL CUDA Graph mode.

Results are shown in the following graph. The input (prompt) length (in tokens) was varied in these experiments across the following values: 500, 1000, 1500, 2000, 4000, 8000, and 16000. The number of warmup iterations and measurement iterations were left at the default values of 10 and 30 respectively.

results

As illustrated in the graph above, this PR improves the performance of the Triton Unified Attention Kernel by approximately 2 times for a batch size of 1 and an input length of 16000 tokens.

Further performance improvements are discussed below: https://github.com/vllm-project/vllm/pull/29020#issuecomment-3562777503

## Test Plan

The unit test `./tests/kernels/attention/test_triton_unified_attention.py` has been updated to reflect the new behavior. It now covers both scenarios for mixed batches:

* Single kernel launch (current approach)
* Separate kernel launches for prefill and decode (new approach for FULL CUDA Graph mode)

Other test suites, such as lm_eval, remain fully compatible and require no modifications.

## Test Result

unit test results for updated Triton unified attention kernel (this PR):
<pre>python3 -m pytest tests/kernels/attention/test_triton_unified_attention.py

================================================ 1536 passed in 229.01s (0:03:49) ================================================
</pre>

<br>
<br>

`lm_eval` results for updated Triton unified attention kernel (this PR) for single kernel launch for mixes batches:
<pre>VLLM_ATTENTION_BACKEND=TRITON_ATTN vllm serve meta-llama/Llama-3.1-8B-Instruct --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}'
lm_eval --model local-completions --model_args base_url=http://localhost:8000/v1/completions,tokenizer=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.798|±  |0.0180|
|     |       |strict-match    |     5|exact_match|↑  |0.786|±  |0.0184|</pre>

`lm_eval` results for updated Triton unified attention kernel (this PR) for separate kernel launches for mixes batches:
<pre>VLLM_ATTENTION_BACKEND=TRITON_ATTN vllm serve meta-llama/Llama-3.1-8B-Instruct --compilation-config '{"cudagraph_mode": "FULL"}'
lm_eval --model local-completions --model_args base_url=http://localhost:8000/v1/completions,tokenizer=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.794|±  |0.0181|
|     |       |strict-match    |     5|exact_match|↑  |0.780|±  |0.0185|</pre>

yields similar `lm_eval` results as FlashAttention:
<pre>lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.794|±  |0.0181|
|     |       |strict-match    |     5|exact_match|↑  |0.772|±  |0.0188|</pre>

@tdoublep @bringlein

---

> [!NOTE]
> No changes detected in the provided diff; no files added, modified, or removed.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 600aaab8d626301f4dbfcd812ccf47b3e8d81ac9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit d285b00607ad3167dfc72fad38d59c4706a04e53. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces separate kernel launches for prefill and decode in mixed batches to enable per-mode tuning and better CUDA Graph capture behavior, while keeping a single unified launch in other modes.
> 
> - Extend `TritonAttentionMetadata`/builder with `num_prefills`, `num_decodes`, `split_launch`, and tuned `seq_threshold_3D`; derive counts via `split_decodes_and_prefills`; add CUDA graph support selection and enable split when `cudagraph_mode=FULL`.
> - Update Triton kernels (`kernel_unified_attention_2d/3d`, `reduce_segments`) with `q_block_offset` and `decode_only` to handle split batches; adjust index search and offsets.
> - Revise `unified_attention` to compute separate BLOCK_Q/BLOCK_M and tile sizes for prefill vs decode, launch 2D for prefills and 2D/3D for decodes based on threshold, and keep backward-compatible defaults.
> - Modify `TritonAttentionImpl.forward` to pass new metadata fields; update unit tests to cover mixed, decode-only, prefill-only cases and both split and unified launch modes.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d285b00607ad3167dfc72fad38d59c4706a04e53. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables separate Triton attention launches for prefill and decode when using FULL CUDA Graphs, allowing per-mode tuning while preserving a single unified launch in other modes.
> 
> - Extend `TritonAttentionMetadata`/builder with `num_prefills`, `num_decodes`, `split_launch`, tuned `seq_threshold_3D`, and softmax buffers; derive counts via `split_decodes_and_prefills`; add `get_cudagraph_support` and enable split when `cudagraph_mode=FULL`.
> - Update `unified_attention` to compute distinct `BLOCK_M/BLOCK_Q` and tile sizes for prefill vs decode; always launch 2D for prefills and choose 2D/3D for decodes by threshold; support split or unified execution with backward-compatible defaults.
> - Modify Triton kernels (`kernel_unified_attention_2d/3d`, `reduce_segments`) to accept `q_block_offset` and `decode_only`, adjusting index calculations for split batches.
> - Pass new fields from `TritonAttentionImpl.forward`; expand unit tests to cover mixed, decode-only, prefill-only batches and both `split_launch` modes.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b6dc91e58c3cf5a3c1db55e0db0765031cc0a9e1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by LucasWilkinson — COMMENTED** [review]
> The changes in this PR are not intended to deliver significant performance gains immediately, but rather to lay the groundwork for future optimizations in upcoming PRs.

IMO its best practice to hold off until we can see the benefits; we can land this as a chain once we know the final PR is delivering perf; this gives reviewers full visibility into the planned changes and ultimate benefit

File: vllm/attention/ops/triton_unified_attention.py [hunk 1]
```diff
@@ -776,25 +796,31 @@ def unified_attention(
...
+        # \sum_i[ceil(query_len[i] / BLOCK_Q)]
+        #   <= \sum_i[floor(query_len[i] / BLOCK_Q) + 1]
+        #    = \sum_i[floor(query_len[i] / BLOCK_Q)] + num_seqs
+        #   <= floor(\sum_i(query_len[i]) / BLOCK_Q) + num_seqs
+        #    = floor(q.shape[0] / BLOCK_Q) + num_seqs
+        total_num_q_blocks = (
+            (q.shape[0] - num_decodes) // BLOCK_Q + num_seqs
+            if split_launch
+            else q.shape[0] // BLOCK_Q + num_seqs
+        )
```
[the remark this answers is not in the document]
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py:1022]: Is addressed by new modification.

File: vllm/attention/ops/triton_unified_attention.py [hunk 2]
```diff
@@ -844,98 +870,220 @@ def unified_attention(
...
+                stride_v_cache_2=v.stride(2),
+                stride_v_cache_3=v.stride(3),
+                query_start_len_ptr=cu_seqlens_q,
+                BLOCK_Q=BLOCK_Q,
+                num_seqs=num_decodes,
+                BLOCK_M=BLOCK_M,
+                q_block_offset=0,
+                decode_only=True,
+                USE_FP8=output_scale is not None,
+            )
```
[the remark this answers is not in the document]
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py]: Is addressed by new modification.

File: vllm/attention/ops/triton_unified_attention.py [hunk 3]
```diff
@@ -844,98 +870,220 @@ def unified_attention(
             BLOCK_Q=BLOCK_Q,
             num_seqs=num_seqs,
             BLOCK_M=BLOCK_M,
+            q_block_offset=num_decodes if split_launch else 0,
+            decode_only=False,
```
[the remark this answers is not in the document]
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py:1078]: That is correct. Speculative decoding is not yet included in the optimizations introduced in this pull request. To handle this properly, reorder_batch_threshold has been set to 1 token, which ensures speculative decode steps are processed by the 2D kernel in the same way as prefills. Furthermore, FULL CUDA Graph mode in combination with speculative decoding is not supported by this pull request, and will be blocked using an assert statement  A follow-up pull request will extend the optimizations in this pull request to also support speculative decoding.

[the remark this answers is not in the document]
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py:1078]: The above mentioned assert statement has been replaced by an override of the `get_cudagraph_support()` class method in the `TritonAttentionMetadataBuilder` class. This override adjusts the `_cudagraph_support` value from `ALWAYS` to `UNIFORM_SINGLE_TOKEN_DECODE` when the combination of FULL CUDA Graph mode and speculative decoding is detected.

**Comment by jvlunteren** [comment]
restore updates in new folder structure

File: vllm/v1/attention/ops/triton_unified_attention.py [hunk 4]
```diff
@@ -1061,28 +1067,157 @@ def unified_attention(
...
+        # Launch the 2D kernel if
+        # 1. No intermediate tiled softmax buffers have been allocated, or
+        # 2. The number of sequences exceeds the configured threshold
+        if (
+            seq_threshold_3D is None
+            or num_par_softmax_segments is None
+            or softmax_segm_output is None
+            or softmax_segm_max is None
+            or softmax_segm_expsum is None
+            or num_seqs > seq_threshold_3D
```
[the remark this answers is not in the document]
Comment (jvlunteren) [reply vllm/v1/attention/ops/triton_unified_attention.py]: Fixed.

File: vllm/v1/attention/backends/triton_attn.py [hunk 5]
```diff
@@ -114,7 +120,7 @@ def mm_prefix_range_tensor(self) -> torch.Tensor | None:
 
 
 class TritonAttentionMetadataBuilder(AttentionMetadataBuilder[TritonAttentionMetadata]):
-    _cudagraph_support: ClassVar[AttentionCGSupport] = AttentionCGSupport.ALWAYS
+    reorder_batch_threshold: int = 1
```
Comment (tdoublep) [inline vllm/v1/attention/backends/triton_attn.py]: Do we really always want to re-order the batch? Shouldn't we do i.f.f. full cuda graph is enabled?

File: vllm/v1/attention/backends/triton_attn.py [hunk 6]
```diff
@@ -144,6 +150,12 @@ def __init__(
             )
         )
 
+        # Check if CUDA Graphs are enabled for prefill.
+        self.prefill_cudagraph_enabled = (
+            self.vllm_config.compilation_config.cudagraph_mode in (CUDAGraphMode.FULL,)
+        )
+        self.split_launch = self.prefill_cudagraph_enabled
```
Comment (tdoublep) [inline vllm/v1/attention/backends/triton_attn.py]: Do we really need both `self.prefill_cudagraph_enabled` and `self.split_launch` ?

File: vllm/v1/attention/backends/triton_attn.py [hunk 7]
```diff
@@ -188,6 +200,31 @@ def __init__(
...
+            speculative_config.num_speculative_tokens
+            if speculative_config is not None
+            else 0
+        )
+
+        # Select the appropriate CUDA graph support mode.
+        if prefill_cudagraph_enabled and (num_spec_tokens > 0):
+            return AttentionCGSupport.UNIFORM_SINGLE_TOKEN_DECODE
+        else:
+            return AttentionCGSupport.ALWAYS
```
Comment (tdoublep) [inline vllm/v1/attention/backends/triton_attn.py]: Ins't this a change in the behaviour of the Triton backend beyond simply separating the kernel launches? What happens on the main branch today when the user tries to combine full cuda graphs with spec decode?

File: vllm/v1/attention/ops/triton_unified_attention.py [hunk 8]
```diff
@@ -876,7 +900,7 @@ def _get_tile_size(
         return 32
 
     # Default behavior
-    if is_prefill:
+    if is_prefill or use_2d_kernel:
```
Comment (tdoublep) [inline vllm/v1/attention/ops/triton_unified_attention.py]: Same question as on the other PR for this change

File: vllm/v1/attention/ops/triton_unified_attention.py [hunk 9]
```diff
@@ -939,53 +966,82 @@ def unified_attention(
     num_queries_per_kv = num_query_heads // num_kv_heads
     head_size = q.shape[2]
 
-    BLOCK_M = (
+    # Assign the following variables if they are not assigned in the attention metadata.
+    # This ensures backward compatibility with callers using an earlier version of this
+    # function. However, it is recommended to include these assignments in the
+    # attention metadata itself, as performing them here may negatively impact
+    # performance.
```
Comment (tdoublep) [inline vllm/v1/attention/ops/triton_unified_attention.py]: Similar feedback to https://github.com/vllm-project/vllm/pull/28576 - I would prefer not to have code here for backward compatibility. Better to change any backend that uses the kernel to add this support.

File: vllm/v1/attention/ops/triton_unified_attention.py [hunk 10]
```diff
@@ -939,53 +966,82 @@ def unified_attention(
...
+    # Assigning Q Block dimensions for prefill and decode.
+    BLOCK_M_2D_PREFILL = (
+        64 if num_queries_per_kv <= 64 else triton.next_power_of_2(num_queries_per_kv)
+    )
+    BLOCK_M_2D_DECODE = (
+        16 if num_queries_per_kv <= 16 else triton.next_power_of_2(num_queries_per_kv)
+    )
+    BLOCK_M_3D_DECODE = (
         16 if num_queries_per_kv <= 16 else triton.next_power_of_2(num_queries_per_kv)
     )
```
Comment (tdoublep) [inline vllm/v1/attention/ops/triton_unified_attention.py]: Couldn't we just use `BLOCK_M_DECODE` (since they are the same)?

File: vllm/v1/attention/backends/triton_attn.py (hunk 5 above)
Comment (jvlunteren) [reply vllm/v1/attention/backends/triton_attn.py]: You are right. I have updated the code accordingly.

File: vllm/v1/attention/backends/triton_attn.py (hunk 6 above)
Comment (jvlunteren) [reply vllm/v1/attention/backends/triton_attn.py]: No, we don't need both. I have adapted the code to remove this redundancy.

File: vllm/v1/attention/backends/triton_attn.py (hunk 7 above)
Comment (jvlunteren) [reply vllm/v1/attention/backends/triton_attn.py]: Yes, this is a change in the behaviour beyond the separate kernel launches. If I recall correctly, I inserted this in order to prevent some problem that occurred, but not related to the separate kernel launches. It should indeed be removed. I will do so.

File: vllm/v1/attention/ops/triton_unified_attention.py (hunk 8 above)
Comment (jvlunteren) [reply vllm/v1/attention/ops/triton_unified_attention.py]: This update actually tries to optimize performance by changing the selection of the tile size. This is not related to the separation of the kernel launches. I will remove the update.

File: vllm/v1/attention/ops/triton_unified_attention.py (hunk 9 above)
Comment (jvlunteren) [reply vllm/v1/attention/ops/triton_unified_attention.py]: I agree. I have updated the code and the backends (tree_attn) that use the kernel.

File: vllm/v1/attention/ops/triton_unified_attention.py (hunk 10 above)
Comment (jvlunteren) [reply vllm/v1/attention/ops/triton_unified_attention.py]: This separation was intended to allow simpler, independent tuning of the decode paths for the 2D and 3D kernels, in case different BLOCK_M values became relevant in the future. Since the values are currently identical, it makes sense to use BLOCK_M_DECODE for both. I will update the code accordingly.
